x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -199,9 +198,12 @@ public synchronized boolean maybeTriggerAsyncJob(long now) {
      * Called at startup after job has been triggered using {@link #maybeTriggerAsyncJob(long)} and the
      * internal state is {@link IndexerState#STARTED}.
      *
+     * Implementors MUST ensure that next.run() gets called.
+     *
      * @param now The current time in milliseconds passed through from {@link #maybeTriggerAsyncJob(long)}
+     * @param next Runnable for the next phase
      */
-    protected abstract void onStartJob(long now);
+    protected abstract void onStart(long now, Runnable next);
```

Comment:
Since we're expecting async calls to be fired in `onStart` we should provide an `ActionListener<Void>` rather than a Runnable ? This way the implementation can fail the job nicely if an error occurs ?